### ISSUE TITLE
Functional multi_head_attention_forward softmax get nan for fp16 mode

### ISSUE THREAD

**Issue by fuzihaofzh** [pytorch/pytorch#60939]
## 🐛 Bug

In torch.nn.functional, the function multi_head_attention_forward has problem in fp16 mode. in line https://github.com/pytorch/pytorch/blob/5b118a7f233a4682e338c9ef14a5ec55183b9a65/torch/nn/functional.py#L4826 , the product of q and k may have inf in the output and it will lead to nan in the following softmax function. Hope this operation can be more rubost in the fp16 mode.
